### ISSUE TITLE
How to load images and text using ImagenTrainer

### ISSUE THREAD

**Issue by korakoe** [lucidrains/imagen-pytorch#175]
Hi, i'm currently trying to find a way to load images and text using ImagenTrainer and the Dataset functions the library provides, i'm a bit confused as there doesnt really seem to be proper documentation on this aside from the unconditioned image generation

**Comment by zhujiajunbryan** [comment]
+1

**Comment by TheChesireCat** [comment]
Take a look at [`elucidated_imagen.py`](https://github.com/lucidrains/imagen-pytorch/blob/main/imagen_pytorch/elucidated_imagen.py) and its [`forward`](https://github.com/lucidrains/imagen-pytorch/blob/main/imagen_pytorch/elucidated_imagen.py#L637) function.

The function accepts as input `images` and `texts`.  The images to be passed must be squares and the text as a list of strings. 

`ElucidatedImagen.forward` is called in [`ImagenTrainer.forward`](https://github.com/lucidrains/imagen-pytorch/blob/main/imagen_pytorch/trainer.py#L902).The function `step_with_dl_iter` is where the `model_input`s are prepared. The keywords being passed to the model (`dl_tuple_output_keywords_names` in `trainer.py`) are important clues which will come handy when designing a custom Dataset class based on `data.py`. Another function you will need when designing your Dataset which returns text embeddings, masks and images as a tuple is the `encode_text` function defined in `trainer.py`

Hope this helps 🌈

**Comment by zhujiajunbryan** [comment]
> Take a look at [`elucidated_imagen.py`](https://github.com/lucidrains/imagen-pytorch/blob/main/imagen_pytorch/elucidated_imagen.py) and its [`forward`](https://github.com/lucidrains/imagen-pytorch/blob/main/imagen_pytorch/elucidated_imagen.py#L637) function.
> 
> The function accepts as input `images` and `texts`. The images to be passed must be squares and the text as a list of strings.
> 
> `ElucidatedImagen.forward` is called in [`ImagenTrainer.forward`](https://github.com/lucidrains/imagen-pytorch/blob/main/imagen_pytorch/trainer.py#L902).The function `step_with_dl_iter` is where the `model_input`s are prepared. The keywords being passed to the model (`dl_tuple_output_keywords_names` in `trainer.py`) are important clues which will come handy when designing a custom Dataset class based on `data.py`. Another function you will need when designing your Dataset which returns text embeddings, masks and images as a tuple is the `encode_text` function defined in `trainer.py`
> 
> Hope this helps 🌈

hi, thanks for the reply, so how should I deal with encoded embeddings? waiting to encode every time or save them in jason files for further use(with a dataset more than 30000 captions)

**Comment by TheChesireCat** [comment]
@lucidrains recommends that you precompute the Embeddings and Masks when you initialize the dataset to save on training time. The t5 encoder can encode items in batches. I'd suggest looking into that.

**Comment by zhujiajunbryan** [comment]
> @lucidrains recommends that you precompute the Embeddings and Masks when you initialize the dataset to save on training time. The t5 encoder can encode items in batches. I'd suggest looking into that.
> 
> Additionally, like you said, you could also save them to a file on disk and reload when you wish to train.

thanks for replying, so sorry really newbie at this, might need babysitting, let's say now I have a dataloader that returns{caption（text）that are in string, tensor image}, what's the interference to convert the caption into encoders without individually decode using T5

**Comment by TheChesireCat** [comment]
Ah yes, so the model can also take as input just the text Embeddings and Masks. So your dataloader can return only the image tensors, mask and embedding tensors without passing a list of raw text. Try reading the [forward](https://github.com/lucidrains/imagen-pytorch/blob/main/imagen_pytorch/elucidated_imagen.py#L637)  function carefully.
